include/smack/smack.h

```diff
@@ -226,11 +524,39 @@ void __SMACK_decls() {
     "ensures (forall q: int :: {$Size[q]} q != p ==> $Size[q] == old($Size[q]));\n"
     "ensures (forall q: int :: {$Alloc[q]} q != p ==> $Alloc[q] == old($Alloc[q]));\n"
     "ensures n >= 0 ==> (forall q: int :: {$base(q)} p <= q && q < p+n ==> $base(q) == p);");
+#endif
 
 #else // NO_REUSE does not reuse previously-allocated addresses
+#ifdef BITVECTOR
+  D("var $Alloc: [ref] bool;");
+  D("var $CurrAddr:ref;");
+  D("procedure $malloc(n: size) returns (p: ref);\n"
+    "modifies $CurrAddr, $Alloc;\n"
+    "ensures $sgt.ref(p, $NULL);\n"
+    "ensures p == old($CurrAddr);\n"
+    "ensures $sgt.ref($CurrAddr, old($CurrAddr));\n"
+    "ensures $sge.ref(n, $NULL) ==> $sge.ref($CurrAddr, $add.ref(old($CurrAddr), n));\n"
+    "ensures $Alloc[p];\n"
+    "ensures (forall q: ref :: {$Alloc[q]} q != p ==> $Alloc[q] == old($Alloc[q]));\n"
+    "ensures $sge.ref(n, $NULL) ==> (forall q: ref :: {$base(q)} $sle.ref(p, q) && $slt.ref(q, $add.ref(p, n)) ==> $base(q) == p);");
+
+  D("procedure $free(p: ref);\n"
+    "modifies $Alloc;\n"
+    "ensures !$Alloc[p];\n"
+    "ensures (forall q: ref :: {$Alloc[q]} q != p ==> $Alloc[q] == old($Alloc[q]));");
+
+  D("procedure $alloca(n: size) returns (p: ref);\n"
+    "modifies $CurrAddr, $Alloc;\n"
+    "ensures $sgt.ref(p, $NULL);\n"
+    "ensures p == old($CurrAddr);\n"
+    "ensures $sgt.ref($CurrAddr, old($CurrAddr));\n"
+    "ensures $sge.ref(n, $NULL) ==> $sge.ref($CurrAddr, $add.ref(old($CurrAddr), n));\n"
+    "ensures $Alloc[p];\n"
+    "ensures (forall q: ref :: {$Alloc[q]} q != p ==> $Alloc[q] == old($Alloc[q]));\n"
+    "ensures $sge.ref(n, $NULL) ==> (forall q: ref :: {$base(q)} $sle.ref(p, q) && $slt.ref(q, $add.ref(p, n)) ==> $base(q) == p);");
+#else
```

Comment:
As noted above, the types `ref` and `size` should replace all uses of `int` for pointer types and size types. The following declarations break this rule.